GUIDELINES.md

```diff
@@ -0,0 +1,153 @@
+
+## Foreword
+
+In advance, we recommend installing the autoformatter [black](https://pypi.org/project/black/) in your IDE. In future we will use `flake8` (or similar) for automatically checking our python codebases. 
+
+Hopefully the conventions laid out here are the easy and intuitive set of [pep8](https://www.python.org/dev/peps/pep-0008/). 
+
+Code reviewers: it is on you to ensure that this style guide has been followed: there are no points for being lenient, but there [non-redeemable] points for being opinionated! We should all feel pressured into making sure that our code meets an acceptable standard.
+
+## `python is not pandas`
+
+tldr; Avoid using `pandas` except for where it makes things ridiculously simple. That is not so say that you should not use `pandas`, but rather that you should generally assume that `pandas` is making your life harder.
+
+Code containing lots of `pandas` operations are almost impossible to review, and therefore have the capacity to accululate vast numbers of bugs.
+
+* In general, `pandas` makes column-wise operations and IO (reading/writing files) dead easy. That said, `pandas` column-wise operations are inherited from `numpy`, and ``numpy` is generally accepted in the place of dataframes.
+* `pandas` is enormous, in many ways. If it can be omitted from your code then you can make big savings in terms of memory usage and requirements clashes, and even CPU time.
+* Instead of Googling how to achieve something in `pandas` with an obscure chaining of functions, break the problem down and solve it yourself. It is highly unlikely that the `pandas` approach to reshaping your data will beat using tools from `numpy`, `itertools`, `functools` and `toolz`, even if you switch to representing data in `numpy` arrays or even as `list` of `dict` (`[{'value': 21}, {'value': 45}]`).
+
+
+## Design patterns
+
+Favour the following design patterns, in this order:
+
+1. Functional programming: using functions to do one thing well, not altering the original data.
+2. Modular code: using functions to eliminate duplication, improve readability and reduce indentation.
+3. Object-oriented programming: Generally avoid, unless you are customising an API (for example `DataFrame`) or defining your own API.
+
+If you are not, at least, adhering to a modular style then you have gone very wrong.
+You should implement unit tests for each of your functions, something which is generally more tricky for object-oriented programming.
+
+## Programming
+
+### Mandatory
+
+* Don't compare boolean values to `True` or `False`.
+* Favour `is not condition` over `not condition is`
+* Don't compare a value to `None` (`value == None`), always favour `value is None`
+
+### Encouraged
+
+* Favour `logging` over `print`
+* Favour using `GLOBAL_VARIABLES` near the top of your code rather than *any* hard-coded variables in your code. These can be more easily factored out into config, and at least puts many of your design decisions in one place.
```

Comment:
What do you mean sorry?

Reply:
That this...
```python

def bar(foo):
  return foo ** 2
```

is usually a better place to start than this...

```python

FOO = 10

def bar():
   return FOO ** 2
```

Unless `FOO` truly is a constant.

Struggling to come up with a good example but potential cases where constant can get murky:
* URL's
* `Path`s

Reply:
I guess I was thinking more:

```python
FOO = 10

def bar(foo):
  return foo ** 2

def main():
    ...
    bar(FOO)
    ...
    bar(FOO)
```

or better still:

```python
FOO = 10

def bar(foo):
  return foo ** 2

def main():
    bar = partial(bar, FOO)
    ...
    bar()
    ...
    bar()
```

Actually I think my wording is quite bad here, perhaps better as:

> * Favour using configuration files, and (faster/lazier/less good/ok) `GLOBAL_VARIABLES` near the top of your code, rather than *repeated* use of hard-coded variables in your code, particularly when with URL and file path variables like `s3://path/to/some/fixed/place`, but also for repeated numeric hyperparameters.